### ISSUE TITLE
W horovod/common/stall_inspector.cc:105] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. 

### ISSUE THREAD

**Issue by ehsab** [horovod/horovod#1890]
**Environment:**
1. Using Horovd docker on Ubuntu18.04 on Google GCP
2. All ssh connections work well between docker containers

I am trying to run pytorch_imagenet_resnet50.py from example folder. I have 4 servers each has 4 V100 GPUs. I run it with the following settings:

1 Server 1 GPU => works fine
1 Server 2 GPUs => works fine
1 Server 4 GPUs => works fine
2 Servers 1 GPU each server => works fine
2 Servers 2 GPUs each server => works fine
4 Servers 1 GPU each server => works fine

2 Servers 4GPUs each server => Doesn't work
4 Servers 2GPUs each server => Doesn't work
4 Servers 4GPUs each server => Doesn't work

As soon as I start using more GPUs it gives me the following error. It seems it has problem handling more than 4 GPUs. I have searched alot on the Internet but couldn't find any solution.

**Bug report:**
Train Epoch     #1:   0%|          | 0/626 [00:00<?, ?it/s][1,0]<stderr>:[2020-04-20 20:29:29. 89665: W horovod/common/stall_inspector.cc:105] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. 
[1,0]<stderr>:Stalled ranks:
[1,0]<stderr>:0: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:1: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:2: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:3: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:4: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:5: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:6: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:7: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:8: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:9: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:10: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:11: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:12: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:13: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:[2020-04-20 20:30:29. 92101: W horovod/common/stall_inspector.cc:105] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. 
[1,0]<stderr>:Stalled ranks:
[1,0]<stderr>:0: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:1: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:2: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:3: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:4: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:5: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:6: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:7: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:8: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:9: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:10: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:11: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:12: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]
[1,0]<stderr>:13: [allreduce.layer2.0.bn2.bias, allreduce.layer2.0.bn2.weight, allreduce.layer2.0.bn3.bias, allreduce.layer2.0.bn3.weight, allreduce.layer2.0.conv3.weight, allreduce.layer2.0.downsample.0.weight ...]

**Comment by tgaddair** [comment]
Hey @ehsab, there's a lot going on in the ResNet script, so my first suggestion would be to try running a script that doesn't require reading any data from disk.  Can you verify that [pytorch_synthetic_benchmark.py](https://github.com/horovod/horovod/blob/master/examples/pytorch_synthetic_benchmark.py) works correctly / exhibits the same behavior when run with > 4 GPUs?

**Comment by ehsab** [comment]
@tgaddair Thanks for the reply. I just tried the one you suggested. Same problem. When I run it on one server with 4 GPUs it works very well and I get the following log:

root@gcp00:/examples# horovodrun --start-timeout 300 -np 4 -H localhost:4,gcp01:0 -p 123  python pytorch_synthetic_benchmark.py
2020-04-21 01:10:55.669032: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libnvinfer.so.6'; dlerror: libnvinfer.so.6: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64
2020-04-21 01:10:55.669130: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libnvinfer_plugin.so.6'; dlerror: libnvinfer_plugin.so.6: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64
2020-04-21 01:10:55.669140: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:30] Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
[1,0]<stderr>:[gcp00:00916] Read -1, expected 11145, errno = 1
[1,0]<stderr>:[gcp00:00916] Read -1, expected 15945, errno = 1
[1,0]<stderr>:[gcp00:00916] Read -1, expected 7433, errno = 1
[1,0]<stderr>:[gcp00:00916] Read -1, expected 19649, errno = 1
[1,0]<stderr>:[gcp00:00916] Read -1, expected 23657, errno = 1
[1,1]<stderr>:[gcp00:00917] Read -1, expected 15208, errno = 1
[1,2]<stderr>:[gcp00:00918] Read -1, expected 15209, errno = 1
[1,3]<stderr>:[gcp00:00919] Read -1, expected 15208, errno = 1
[1,0]<stdout>:Model: resnet50
[1,0]<stdout>:Batch size: 32
[1,0]<stdout>:Number of GPUs: 4
[1,0]<stdout>:Running warmup...
[1,0]<stdout>:Running benchmark...
[1,0]<stdout>:Iter #0: 261.3 img/sec per GPU
[1,0]<stdout>:Iter #1: 262.9 img/sec per GPU
[1,0]<stdout>:Iter #2: 264.8 img/sec per GPU
[1,0]<stdout>:Iter #3: 262.1 img/sec per GPU
[1,0]<stdout>:Iter #4: 265.1 img/sec per GPU
[1,0]<stdout>:Iter #5: 262.7 img/sec per GPU
[1,0]<stdout>:Iter #6: 263.7 img/sec per GPU
[1,0]<stdout>:Iter #7: 263.4 img/sec per GPU
[1,0]<stdout>:Iter #8: 266.2 img/sec per GPU
[1,0]<stdout>:Iter #9: 262.8 img/sec per GPU
[1,0]<stdout>:Img/sec per GPU: 263.5 +-2.7
[1,0]<stdout>:Total img/sec on 4 GPU(s): 1054.0 +-11.0

However, When I run it on 2 servers each has 4 GPUs it stalls:

root@gcp00:/examples# horovodrun --start-timeout 300 -np 8 -H localhost:4,gcp01:4 -p 123  python pytorch_synthetic_benchmark.py
2020-04-21 01:11:30.073254: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libnvinfer.so.6'; dlerror: libnvinfer.so.6: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64
2020-04-21 01:11:30.073334: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libnvinfer_plugin.so.6'; dlerror: libnvinfer_plugin.so.6: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64
2020-04-21 01:11:30.073349: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:30] Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
[1,0]<stderr>:[gcp00:01137] Read -1, expected 27041, errno = 1
[1,0]<stderr>:[gcp00:01137] Read -1, expected 12489, errno = 1
[1,0]<stderr>:[gcp00:01137] Read -1, expected 14593, errno = 1
[1,0]<stderr>:[gcp00:01137] Read -1, expected 16713, errno = 1
[1,0]<stderr>:[gcp00:01137] Read -1, expected 10369, errno = 1
[1,1]<stderr>:[gcp00:01138] Read -1, expected 11116, errno = 1
[1,2]<stderr>:[gcp00:01139] Read -1, expected 11117, errno = 1
[1,6]<stderr>:[gcp01:00782] Read -1, expected 11117, errno = 1
[1,5]<stderr>:[gcp01:00781] Read -1, expected 11116, errno = 1
[1,1]<stderr>:[gcp00:01138] Read -1, expected 8856, errno = 1
[1,2]<stderr>:[gcp00:01139] Read -1, expected 8857, errno = 1
[1,6]<stderr>:[gcp01:00782] Read -1, expected 8857, errno = 1
[1,5]<stderr>:[gcp01:00781] Read -1, expected 8856, errno = 1
[1,0]<stdout>:Model: resnet50
[1,0]<stdout>:Batch size: 32
[1,0]<stdout>:Number of GPUs: 8
[1,0]<stdout>:Running warmup...

**Comment by prudhvee** [comment]
Hi @ehsab,
I am also facing similar issue. Did you manage to root cause/solve this issue?
On 8GPU machine I am facing the issue if run my model training using 8threads `-np 8`.
However, on the same 8GPU machine if I run training on 4threads `-np 4`, it works fine.
